File: Assets/Script/Example/Body.lua
-- [yue]: Script/Example/Body.yue
local Vec2 = Dora.Vec2 -- 1
local BodyDef = Dora.BodyDef -- 1
local PhysicsWorld = Dora.PhysicsWorld -- 1
local Body = Dora.Body -- 1
local threadLoop = Dora.threadLoop -- 1
local App = Dora.App -- 1
local ImGui = Dora.ImGui -- 1
local gravity <const> = Vec2(0, -10) -- 3
local groupZero <const> = 0 -- 5
local groupOne <const> = 1 -- 6
local groupTwo <const> = 2 -- 7
local terrainDef -- 9
do -- 9
	local _with_0 = BodyDef() -- 9
	_with_0.type = "Static" -- 10
	_with_0:attachPolygon(800, 10, 1, 0.8, 0.2) -- 11
	terrainDef = _with_0 -- 9
end -- 9
local polygonDef -- 13
do -- 13
	local _with_0 = BodyDef() -- 13
	_with_0.type = "Dynamic" -- 14
	_with_0.linearAcceleration = gravity -- 15
	_with_0:attachPolygon({ -- 17
		Vec2(60, 0), -- 17
		Vec2(30, -30), -- 18
		Vec2(-30, -30), -- 19
		Vec2(-60, 0), -- 20
		Vec2(-30, 30), -- 21
		Vec2(30, 30) -- 22
	}, 1, 0.4, 0.4) -- 16
	polygonDef = _with_0 -- 13
end -- 13
local diskDef -- 25
do -- 25
	local _with_0 = BodyDef() -- 25
	_with_0.type = "Dynamic" -- 26
	_with_0.linearAcceleration = gravity -- 27
	_with_0:attachDisk(60, 1, 0.4, 0.4) -- 28
	diskDef = _with_0 -- 25
end -- 25
do -- 30
	local world = PhysicsWorld() -- 30
	world.y = -200 -- 31
	world.showDebug = true -- 32
	world:setShouldContact(groupZero, groupOne, false) -- 34
	world:setShouldContact(groupZero, groupTwo, true) -- 35
	world:setShouldContact(groupOne, groupTwo, true) -- 36
	world:addChild((function() -- 38
		local _with_0 = Body(terrainDef, world, Vec2.zero) -- 38
		_with_0.group = groupTwo -- 39
		return _with_0 -- 38
	end)()) -- 38
	world:addChild((function() -- 41
		local _with_0 = Body(polygonDef, world, Vec2(0, 500), 15) -- 41
		_with_0.group = groupOne -- 42
		return _with_0 -- 41
	end)()) -- 41
	world:addChild((function() -- 44
		local _with_0 = Body(diskDef, world, Vec2(50, 800)) -- 44
		_with_0.group = groupZero -- 45
		_with_0.angularRate = 90 -- 46
		return _with_0 -- 44
	end)()) -- 44
end -- 30
local windowFlags = { -- 51
	"NoDecoration", -- 51
	"AlwaysAutoResize", -- 51
	"NoSavedSettings", -- 51
	"NoFocusOnAppearing", -- 51
	"NoNav", -- 51
	"NoMove" -- 51
} -- 51
return threadLoop(function() -- 59
	local width -- 60
	width = App.visualSize.width -- 60
	ImGui.SetNextWindowBgAlpha(0.35) -- 61
	ImGui.SetNextWindowPos(Vec2(width - 10, 10), "Always", Vec2(1, 0)) -- 62
	ImGui.SetNextWindowSize(Vec2(240, 0), "FirstUseEver") -- 63
	return ImGui.Begin("Body", windowFlags, function() -- 64
		ImGui.Text("Body (Yuescript)") -- 65
		ImGui.Separator() -- 66
		return ImGui.TextWrapped("Basic usage to create physics bodies!") -- 67
	end) -- 67
end) -- 67

File: Tools/dora-rust/dora-test/src/tests/body.rs
/* Copyright (c) 2016-2026 Li Jin <dragon-fly@qq.com>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE. */

use dora_ssr::*;

pub fn test() {
	let gravity = Vec2::new(0.0, -10.0);
	let group_zero = 0;
	let group_one = 1;
	let group_two = 2;

	let mut terrain_def = BodyDef::new();
	terrain_def.set_type(BodyType::Static);
	terrain_def.attach_polygon(800.0, 10.0, 1.0, 0.8, 0.2);

	let mut polygon_def = BodyDef::new();
	polygon_def.set_type(BodyType::Dynamic);
	polygon_def.set_linear_acceleration(&gravity);
	polygon_def.attach_polygon_with_vertices(&vec![
		Vec2::new(60.0, 0.0),
		Vec2::new(30.0, -30.0),
		Vec2::new(-30.0, -30.0),
		Vec2::new(-60.0, 0.0),
		Vec2::new(-30.0, 30.0),
		Vec2::new(30.0, 30.0)
	], 1.0, 0.4, 0.4);

	let mut disk_def = BodyDef::new();
	disk_def.set_type(BodyType::Dynamic);
	disk_def.set_linear_acceleration(&gravity);
	disk_def.attach_disk(60.0, 1.0, 0.4, 0.4);

	let mut world = PhysicsWorld::new();
	world.set_y(-200.0);
	world.set_should_contact(group_zero, group_one, false);
	world.set_should_contact(group_zero, group_two, true);
	world.set_should_contact(group_one, group_two, true);
	world.set_show_debug(true);

	let mut body = Body::new(&terrain_def, &world, &Vec2::zero(), 0.0);
	body.set_group(group_two);
	world.add_child(&body);

	let mut body_p = Body::new(&polygon_def, &world, &Vec2::new(0.0, 500.0), 15.0);
	body_p.set_group(group_one);
	world.add_child(&body_p);

	let mut body_d = Body::new(&disk_def, &world, &Vec2::new(50.0, 800.0), 0.0);
	body_d.set_group(group_zero);
	body_d.set_angular_rate(90.0);
	world.add_child(&body_d);

	let windows_flags =
		ImGuiWindowFlag::NO_DECORATION |
		ImGuiWindowFlag::AlwaysAutoResize |
		ImGuiWindowFlag::NoSavedSettings |
		ImGuiWindowFlag::NoFocusOnAppearing |
		ImGuiWindowFlag::NO_NAV |
		ImGuiWindowFlag::NoMove;
	let mut imgui_node = Node::new();
	imgui_node.schedule(Box::new(move |_| {
		let width = App::get_visual_size().width;
		ImGui::set_next_window_bg_alpha(0.35);
		ImGui::set_next_window_pos_opts(&Vec2::new(width - 10.0, 10.0), ImGuiCond::Always, &Vec2::new(1.0, 0.0));
		ImGui::set_next_window_size_opts(&Vec2::new(240.0, 0.0), ImGuiCond::FirstUseEver);
		ImGui::begin_opts("Body", windows_flags, || {
			ImGui::text("Body (Rust)");
			ImGui::separator();
			ImGui::text_wrapped("Basic usage to create physics bodies!");
		});
		false
	}));
}
